---
title: "BMEG310 BRCA Report II"
author: "Group 5: Calla Stanley, Hemanya Sharma, Evan Alanen, Nikki Comfort"
date: ""
output: html_document
---

```{r}
install.packages("glmnet")

```


```{r setup, include=FALSE}

library("TCGAbiolinks")
library("SummarizedExperiment")
library(survival)
library(survminer)
library(dplyr)
library(ggplot2)
library(patchwork)
library(pheatmap)
library(reshape2)
library(devtools)
library(remotes)
library(ggbiplot)
library(org.Hs.eg.db)
library(DESeq2)

#install.packages("glmnet")
library(stringr)
library(biomaRt)
library(scales)
library(glmnet)
library(pROC)
library(ggrepel)

# add any other libraries used
```

```{r Load Files}
setwd("C:/Users/heman/OneDrive - UBC/Desktop/Intro-to-R-BMEG310")
clinical <- read.csv("TCGA-BRCA/data_clinical_patient.txt", sep = "\t", header = TRUE, skip = 4)
mutations <- read.csv("TCGA-BRCA/data_mutations.txt", sep = "\t", header = TRUE)
rnaseq <- read.csv("TCGA-BRCA/RNAseq_BRCA.csv", header = TRUE)
```


# --- Investigation 1: Survival of Mestastasis in RNA ---

```{r Investigation 1, fig.width=25, fig.height=8, Figure 1.1}
gene_ids <- rnaseq[[1]]
expr_mat <- as.matrix(rnaseq[, -1])
rownames(expr_mat) <- gene_ids

gene_vars <- rowVars(expr_mat, na.rm = TRUE)
top_n <- 1000  ### find significance?
top_genes <- names(sort(gene_vars, decreasing = TRUE))[1:top_n]
expr_top <- expr_mat[top_genes, , drop = FALSE]

expr_log <- log2(expr_top + 1)       
 transform
expr_scaled <- t(scale(t(expr_log)))       

## Ask if we find unique patients or use all of the info
sample_ids  <- colnames(expr_scaled)
patient_ids <- substr(sample_ids, 1, 12)
unique_patients <- unique(patient_ids)

expr_patient <- sapply(unique_patients, function(pid){
  cols <- which(patient_ids == pid)
  if (length(cols) == 1) {
    expr_scaled[, cols]
  } else {
    rowMeans(expr_scaled[, cols, drop = FALSE])
  }
})

expr_patient <- as.matrix(expr_patient)
colnames(expr_patient) <- unique_patients

dist_pat <- dist(t(expr_patient), method = "euclidean")
hc <- hclust(dist_pat, method = "ward.D2")

k <- 3   # try more later
clusters <- cutree(hc, k = k)

cluster_df <- data.frame(
  patient_id = names(clusters),
  rna_cluster = factor(clusters)
)

plot(hc, labels = FALSE,
     main = sprintf("Hierarchical Clustering of Patients (k=%d)", k),
     xlab = "",
     sub = "",
     cex.main = 2,  # main title size
     cex.lab = 1.7    # axis label size
)

```
## Figure 1.1, Clustering of patients and Gene Expression from RNA data


```{r Investigation 1 Continued}
annotation <- data.frame(Cluster = factor(clusters))
rownames(annotation) <- names(clusters)

pheatmap(expr_patient,
         show_rownames = FALSE,
         show_colnames = FALSE,
         annotation_col = annotation,
         clustering_method = "ward.D2",
         main = "Top 1000 Most Variable Genes")

head(cluster_df)
cat("Clusters created:", length(unique(cluster_df$rna_cluster)), "\n")
cat("Patients clustered:", ncol(expr_patient), "\n")
```


```{r Investigation 1 Continued, Figure 1.2}

expr_all <- as.matrix(rnaseq[, -1])
rownames(expr_all) <- rnaseq[[1]]

# log2 transform
expr_all_log <- log2(expr_all + 1)

# z-score scale by gene
expr_all_scaled <- t(scale(t(expr_all_log)))

sample_ids_all  <- colnames(expr_all_scaled)
patient_ids_all <- substr(sample_ids_all, 1, 12)
unique_patients_all <- unique(patient_ids_all)

# Aggregate replicates per patient
expr_patient_all <- sapply(unique_patients_all, function(pid) {
  idx <- which(patient_ids_all == pid)
  if (length(idx) == 1) expr_all_scaled[, idx]
  else rowMeans(expr_all_scaled[, idx, drop = FALSE])
})

expr_patient_all <- as.matrix(expr_patient_all)
colnames(expr_patient_all) <- unique_patients_all

spp1_id <- "ENSG00000118785.14"

spp1_raw <- expr_all_scaled[spp1_id, , drop = FALSE]

# Convert to patient-level aggregated expression
spp1_patient <- sapply(unique_patients_all, function(pid) {
  idx <- which(patient_ids_all == pid)
  if (length(idx) == 1) spp1_raw[, idx]
  else mean(spp1_raw[, idx])
})

# Construct dataframe aligned with clusters
spp1_df <- data.frame(
  patient_id = names(spp1_patient),
  SPP1 = as.numeric(spp1_patient)
) %>%
  inner_join(cluster_df, by = "patient_id")

# --- Plot SPP1 Expression by RNA Cluster ---
ggplot(spp1_df, aes(x = rna_cluster, y = SPP1, fill = rna_cluster)) +
  geom_boxplot() +
  theme_classic() +
  labs(
    title = "SPP1 Expression by RNA Cluster",
    x = "RNA Cluster",
    y = "SPP1 Expression (scaled log2)"
  ) +
  scale_fill_brewer(palette = "Dark2") +
  theme(
    plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
    axis.title.x = element_text(size = 16, face = "bold"),
    axis.title.y = element_text(size = 16, face = "bold"),
    axis.text = element_text(size = 14),
    legend.position = "none"
  )

spp1_cluster_means <- spp1_df %>%
  dplyr::group_by(rna_cluster) %>%
  dplyr::summarise(mean_SPP1 = mean(SPP1, na.rm = TRUE))

spp1_cluster_means
```

## Figure 1.2, Prevalence of SPP1 Marker in Clusters


```{r Investigation 1 Survival, Figure 1.3}
clinical$patient_id <- clinical$PATIENT_ID

# --- Fix patient_id in clusters ---
cluster_df$patient_id <- substr(cluster_df$patient_id, 1, 12)
cluster_df$patient_id <- gsub("\\.", "-", cluster_df$patient_id)

# --- Ensure OS event and months ---
clinical$OS_EVENT <- ifelse(grepl("DECEASED", clinical$OS_STATUS, ignore.case = TRUE), 1, 0)
clinical$OS_MONTHS <- as.numeric(clinical$OS_MONTHS)

# --- Join clinical + cluster ---
surv_df <- dplyr::inner_join(clinical, cluster_df, by = "patient_id")

# --- Check ---
nrow(surv_df)

# --- Survival model ---
fit <- survfit(Surv(OS_MONTHS, OS_EVENT) ~ rna_cluster, data = surv_df)

ggsurvplot(
  fit,
  data = surv_df,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.height = 0.35,
  xlab = "Overall Survival (Months)",
  ylab = "Survival Probability",
  legend.title = "RNA Cluster",
  palette = "Dark2",
  title = "Overall Survival by RNA Expression Clusters"
)
```
## Figure 1.3, Survival by RNA Clusters 

# --- Investigation 2: Tumor Stage Impacts on Gene Expression ---


#Data Normalization:
```{r data normalization}
# Configure raw rnaseq data:
gene_colname <- colnames(rnaseq)[1]

gene_ids2 <- rnaseq[[gene_colname]]

count_mat <- rnaseq[, -1]
count_mat <- as.matrix(count_mat)

rownames(count_mat) <- gene_ids2

# DESeq2 Dataset:
sample_info <- data.frame(row.names = colnames(count_mat),
                          condition = rep("BRCA", ncol(count_mat)))
```


#Clean Tumor Stage Data
```{r Cleaning Tumor Stage Data}
stage <- tolower(clinical$AJCC_PATHOLOGIC_TUMOR_STAGE)
stage <- gsub("\\s*[abc]$", "", stage)

#remove the end stuff (subtypes)
stage[stage %in% c("", "stage x")] <- NA

# remove the word "stage "
stage <- gsub("stage ", "", stage)

convert_roman <- function(x) {
  if (is.na(x)) return(NA_integer_)
  
  if (x == "i")  return(1L)
  if (x == "ii") return(2L)
  if (x == "iii") return(3L)
  if (x == "iv") return(4L)
  
  return(NA_integer_)
}


clinical$STAGE_NUMERIC <- sapply(stage, convert_roman)

clinical$STAGE_FACTOR <- factor(clinical$STAGE_NUMERIC,
                                levels = c(1,2,3,4),
                                labels = c("Stage I", "Stage II", "Stage III", 
                                           "Stage IV"))

```

```{r Finding Number of Indiviuals in Advanced vs Early Stage}
clinical$TUMOR_STAGE <- ifelse(
  clinical$STAGE_NUMERIC %in% c(1,2), "Early",
  ifelse(clinical$STAGE_NUMERIC %in% c(3,4), "Advanced", NA)
)

table(clinical$TUMOR_STAGE, useNA = "ifany")
#note: clinical$TUMOR_STAGE tells early vs advanced and NA has already been removed.

```
```{r Figure 2.1}
ggplot(subset(clinical, !is.na(TUMOR_STAGE)), 
       aes(x = TUMOR_STAGE)) +
  geom_bar(fill = "hotpink") +
  labs(
    title = "Counts of Early vs Advanced Tumor Stage",
    x = "Tumor Stage Group",
    y = "Number of Patients"
  )


```
## Figure 2.1, Counts for Early vs Advanced Stage


```{r Differential Analysis Part 1}
# -- Finding Common Patients btwn clinical + rnaseq --
# RNA: sample IDs from count matrix
rna_samples       <- colnames(count_mat)
rna_samples_clean <- gsub("\\.", "-", rna_samples)
rna_patient_ids   <- substr(rna_samples_clean, 1, 12)

#head(rna_patient_ids)

# Clinical: patient IDs
clinical_patient_ids <- substr(clinical$PATIENT_ID, 1, 12)
#head(clinical_patient_ids)

# Common patients
common_patients <- intersect(rna_patient_ids, clinical_patient_ids)

#head(common_patients)
print("Number of common patients between clinical and RNASeq:")
length(common_patients)

```
```{r Differential Analysis Part 2}
# -- Clean Data + Make Metadata --
clinical$PATIENT_12 <- substr(clinical$PATIENT_ID, 1, 12)

# keep matching rows
clin_meta <- clinical[clinical$PATIENT_12 %in% common_patients, ]

# remove duplicates
clin_meta <- clin_meta[!duplicated(clin_meta$PATIENT_12), ]

# reorder to match RNA order
order_idx <- match(rna_patient_ids, clin_meta$PATIENT_12)
clin_meta <- clin_meta[order_idx, ]

# drop NA tumor stage after alignment
valid <- !is.na(clin_meta$TUMOR_STAGE)
clin_meta <- clin_meta[valid, ]
count_mat_final <- count_mat[, valid]

# build metadata
metadata_df <- data.frame(
  row.names = colnames(count_mat_final),
  Stage = factor(clin_meta$TUMOR_STAGE, levels = c("Early", "Advanced"))
)

```

```{r Differential Analysis Part 3}
# -- Perform DESeq --
dds <- DESeqDataSetFromMatrix(
    countData = count_mat_final,
    colData   = metadata_df,
    design    = ~ Stage
)

dds <- dds[rowSums(counts(dds)) > 0, ]
dds <- DESeq(dds)

```


```{r Differential Analysis Part 4, Table 2.1}
res <- results(dds, contrast = c("Stage", "Advanced", "Early"))
res_shrunk <- lfcShrink(dds, coef = "Stage_Advanced_vs_Early", type = "apeglm")
res_ordered <- res_shrunk[order(res_shrunk$pvalue), ]
head(res_ordered)

```
## Table 2.1, P-Value Ordered Results for DESeq

```{r volcano plot setup}
library(ggplot2)
library(ggrepel)

# Function 1: Prepare DESeq2 results for plotting
prepare_res_for_plot <- function(res) {
  res_df <- as.data.frame(res)
  res_df$gene <- rownames(res_df)
  return(res_df)
}

# Prepare the results for the plot (use res or res_shrunk)
res_df <- prepare_res_for_plot(res_shrunk)

# Function 2: Volcano plot
volcplot <- function(data, padj_threshold = 0.05, log2Fold_threshold = 1,
                     plot_title = 'Volcano Plot', plot_subtitle = NULL) {
  
  df <- data   # local copy so original isn't modified
  
  # Set FC thresholds
  neg_log2fc <- -log2Fold_threshold
  pos_log2fc <- log2Fold_threshold
  
  # Replace NA values
  df$padj[is.na(df$padj)] <- 1
  df$log2FoldChange[is.na(df$log2FoldChange)] <- 0
  
  # Label genes
  df$log2fc_threshold <- ifelse(
    df$log2FoldChange >= pos_log2fc & df$padj <= padj_threshold, 'up',
    ifelse(df$log2FoldChange <= neg_log2fc & df$padj <= padj_threshold, 'down', 'ns')
  )
  
  # Count categories
  up_genes <- sum(df$log2fc_threshold == 'up')
  down_genes <- sum(df$log2fc_threshold == 'down')
  unchanged_genes <- sum(df$log2fc_threshold == 'ns')
  
  legend_labels <- c(
    paste0('Up: ', up_genes),
    paste0('Not significant: ', unchanged_genes),
    paste0('Down: ', down_genes)
  )
  
  x_axis_limits <- ceiling(max(abs(df$log2FoldChange)))
  
  plot_colors <- c(
    'up' = 'springgreen',
    'ns' = 'gray',
    'down' = 'cornflowerblue'
  )
  
  ggplot(df) +
    geom_point(aes(x = log2FoldChange, y = -log10(padj), color = log2fc_threshold),
               alpha = 0.25, size = 1.5) +
    geom_vline(xintercept = c(neg_log2fc, pos_log2fc), linetype = 'dashed') +
    geom_hline(yintercept = -log10(padj_threshold), linetype = 'dashed') +
    scale_x_continuous('log2(FC)', limits = c(-x_axis_limits, x_axis_limits)) +
    scale_color_manual(values = plot_colors, labels = legend_labels) +
    labs(
      color = paste('log2Fold:', log2Fold_threshold, ', padj â‰¤', padj_threshold),
      title = plot_title,
      subtitle = plot_subtitle
    ) +
    theme_bw(base_size = 24) +
    theme(
      aspect.ratio = 1,
      axis.text = element_text(color = 'black'),
      legend.margin = margin(0, 0, 0, 0),
      legend.box.margin = margin(0, 0, 0, 0),
      legend.spacing.x = unit(0.2, 'cm')
    )
}


```


```{r, fig.width=7, fig.height=6 Figure 2.2}
# Generate the volcano plot
volcplot(
  data = res_df,
  padj_threshold = 0.05,
  log2Fold_threshold = 1,
  plot_title = "Volcano Plot for DESeq2 Results",
  plot_subtitle = "Comparison: Advanced vs Early Stage Tumors"
)
```
## Figure 2.2, Volcano Plot for Tumor Stage


```{r Figure 2.3}
library(DESeq2)
library(pheatmap)

# Apply VST transformation once (fastest for large cohorts)
vsd <- vst(dds, blind = FALSE)
mat <- assay(vsd)

sampleDists <- dist(t(mat))
sampleDistMatrix <- as.matrix(sampleDists)

rownames(sampleDistMatrix) <- colnames(count_mat_final)
colnames(sampleDistMatrix) <- colnames(count_mat_final)

annot_df <- data.frame(Stage = clin_meta$TUMOR_STAGE)
rownames(annot_df) <- colnames(count_mat_final)

# Plot heatmap 
pheatmap(
  sampleDistMatrix,
  clustering_distance_rows = sampleDists,
  clustering_distance_cols = sampleDists,
  show_rownames = FALSE,
  show_colnames = FALSE,
  annotation_col = annot_df,
  main = "Sample-to-Sample Distance Heatmap (VST-transformed)"
)


```
# Figure 2.3, Heatmap for Tumor Stage

```{r PCA, Figure 2.5}
top_genes <- head(rownames(res_ordered), 100)
mat <- assay(vsd)[top_genes, ]

pca <- prcomp(t(mat))

df <- data.frame(
  PC1 = pca$x[,1],
  PC2 = pca$x[,2],
  Stage = metadata_df$Stage
)

ggplot(df, aes(PC1, PC2, color = Stage)) +
  geom_point(size=3) +
  theme_minimal()

```
# Figure 2.5, PCA for Tumor Stage


```{r Logistic Regression, Table 2.2}
# pick top 10 strongest DE genes
top10 <- rownames(res_ordered)[1:10]

# extract their VST expression
mat10 <- t(assay(vsd)[top10, ])

# build dataframe with response + predictors
logit_df <- data.frame(
  Stage = factor(metadata_df$Stage, levels = c("Early", "Advanced")),
  mat10
)

#logistic regression
glm.fit <- glm(Stage ~ ., data = logit_df, family = binomial)
summary(glm.fit)

glm.probs <- predict(glm.fit, type = "response")
#head(glm.probs)

glm.pred <- ifelse(glm.probs > 0.5, "Advanced", "Early")
glm.pred <- factor(glm.pred, levels = c("Early", "Advanced"))

table(Predicted = glm.pred, Actual = logit_df$Stage)
mean(glm.pred == logit_df$Stage)

```
## Table  2.2, Logistic Regression


# --- Investigation 3: ____ ---



# --- Investigation 4: Gene expression biomarkers predicting tumor recurrence
```{r data and recurrence setup}
#names(clinical)
#head(clinical)

clinical <- clinical %>%
  mutate(
    PatientID = toupper(PATIENT_ID)
  )
#table(clinical$NEW_TUMOR_EVENT_AFTER_INITIAL_TREATMENT, useNA = "ifany")
#table(clinical$DFS_STATUS, useNA = "ifany")

clinical <- clinical %>%
  mutate(
    Recurrence = case_when(
      # Any explicit new tumor event yes
      NEW_TUMOR_EVENT_AFTER_INITIAL_TREATMENT == "YES" ~ "Recurrence",
      NEW_TUMOR_EVENT_AFTER_INITIAL_TREATMENT == "NO"  ~ "NoRecurrence",
      
      !is.na(DFS_STATUS) & str_detect(DFS_STATUS, regex("1:Recurred/Progressed", 
                                                        ignore_case = TRUE)) ~ "Recurrence",
      !is.na(DFS_STATUS) & str_detect(DFS_STATUS, regex("0:DiseaseFree", 
                                                        ignore_case = TRUE)) ~ "NoRecurrence",
      TRUE ~ NA_character_
    ),
    Recurrence = factor(Recurrence, levels = c("NoRecurrence", "Recurrence"))
  )
#table(clinical$Recurrence, useNA = "ifany")

rnaseq_raw <- rnaseq
colnames(rnaseq_raw) <- gsub("\\.", "-", colnames(rnaseq_raw))
gene_col <- colnames(rnaseq_raw)[1]

expr_df <- rnaseq_raw
rownames(expr_df) <- expr_df[[gene_col]]
expr_df[[gene_col]] <- NULL

# Configure patient IDs
sample_ids_full <- colnames(expr_df)
patient_ids <- substr(sample_ids_full, 1, 12)

expr_samples_x_genes <- t(as.matrix(expr_df))
expr_collapsed <- rowsum(expr_samples_x_genes, group = patient_ids)  # rows = patients
expr_mat <- t(expr_collapsed)
colnames(expr_mat) <- rownames(expr_collapsed)

# Align with clinical:
clin_rec <- clinical %>% filter(!is.na(Recurrence))
common_ids <- intersect(clin_rec$PatientID, colnames(expr_mat))
length(common_ids)

clin_rec <- clin_rec %>% filter(PatientID %in% common_ids)
expr_sub <- expr_mat[, common_ids]
expr_sub <- expr_sub[, clin_rec$PatientID] 
```

```{r Linear model DE}
# Differential Expression: Recurrance vs NoRecurrance
# Linear modeling (lm) per gene:
group <- as.factor(clin_rec$Recurrence)

lm_test <- function(y_gene) {
  fit <- lm(y_gene ~ group)
  summary(fit)$coefficients["groupRecurrence", c("Estimate","Pr(>|t|)")]
}

lm_res <- t(apply(expr_sub, 1, lm_test))
lm_res <- as.data.frame(lm_res)
colnames(lm_res) <- c("logFC","pvalue")
lm_res$pvalue[is.na(lm_res$pvalue)] <- 1
lm_res$padj <- p.adjust(lm_res$pvalue, method = "fdr")

sum(lm_res$padj < 0.05)
head(lm_res[order(lm_res$padj), ])

lm_res$ensembl_clean <- sub("\\..*$", "", rownames(lm_res))

# Map genes symbols:
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
mapping <- getBM(
  attributes = c("ensembl_gene_id", "hgnc_symbol"),
  filters = "ensembl_gene_id",
  values  = lm_res$ensembl_clean,
  mart = mart
)

# Remove extremely low expression genes (keep >1 TPM-equivalent):
gene_means <- rowMeans(expr_sub, na.rm = TRUE)
keep <- gene_means > 1
lm_res_filt <- lm_res[keep, ]

sum(lm_res_filt$padj < 0.05)
head(lm_res_filt[order(lm_res_filt$padj), ])
```

```{r DESeq DE Analysis}
#head(expr_mat[,1:5])
dds <- DESeqDataSetFromMatrix(
  countData = round(expr_sub),      
  colData   = data.frame(Recurrence = clin_rec$Recurrence),
  design    = ~ Recurrence
)
dds <- DESeq(dds)

res_deseq <- results(dds, contrast = c("Recurrence", "Recurrence", "NoRecurrence"))
res_deseq$gene <- rownames(res_deseq)
res_deseq <- res_deseq[order(res_deseq$padj), ]
head(res_deseq)
sum(res_deseq$padj < 0.05, na.rm = TRUE) # number of significant genes
```

```{r Figure 4.1 DE volcano plot}
ggplot(as.data.frame(res_deseq), aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  labs(
    title = "Figure 4.1 DESeq2 Differential Expression: Recurrence vs No Recurrence",
    x = "log2 Fold Change",
    y = "-log10(p-value)"
  )
```
## Figure 4.1 DE volcano plot

```{r Significant Biomarkers}
#head(res_deseq, 50)
res50 <- res_deseq[order(res_deseq$padj), ][1:50, ]
res50$ensembl_clean <- sub("\\..*", "", rownames(res50)) # remove version numbers

# Annotate top 50 and merge:
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
annot <- getBM(
  attributes = c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
  filters    = "ensembl_gene_id",
  values     = res50$ensembl_clean,
  mart       = mart
)

res50_annot <- merge(
  as.data.frame(res50),
  annot,
  by.x = "ensembl_clean",
  by.y = "ensembl_gene_id",
  all.x = TRUE
)
#res50_annot

# Create meaningful gene identifiers:
res50_annot$label <- ifelse(
  res50_annot$hgnc_symbol == "" | is.na(res50_annot$hgnc_symbol),
  paste0(res50_annot$gene_biotype, " (", res50_annot$gene, ")"),
  res50_annot$hgnc_symbol
)

# Select top N:
N <- 20
valid <- res50_annot[!is.na(res50_annot$label) & res50_annot$label != "", ]
topN <- head(valid, N)
topN[, c("label", "log2FoldChange", "padj", "gene_biotype")]
```

```{r Table 4.1 Biomarkers}
biomarker_table <- topN %>%
  select(
    Gene = label,
    log2FoldChange,
    padj,
    Biotype = gene_biotype,
    ENSEMBL = ensembl_clean
  ) %>%
  mutate(
    log2FoldChange = round(log2FoldChange, 3),
    padj = signif(padj, 3)
  ) %>%
  arrange(padj)

biomarker_table
```


```{r Figure 4.2 Heatmap top20}
norm_counts <- counts(dds, normalized=TRUE)
rownames(norm_counts) <- sub("\\..*", "", rownames(norm_counts))

mat_top20 <- norm_counts[topN$ensembl_clean, clin_rec$PatientID]
mat_top20 <- mat_top20[, clin_rec$PatientID]
#topN$ensembl_clean[!topN$ensembl_clean %in% rownames(norm_counts)]
#table(colnames(norm_counts) %in% clin_rec$PatientID)

ann <- data.frame(Recurrence = clin_rec$Recurrence)
rownames(ann) <- clin_rec$PatientID
#all(colnames(mat_top20) == rownames(ann))
ann <- ann[colnames(mat_top20), , drop=FALSE]
ann$Recurrence <- droplevels(as.factor(ann$Recurrence))

annotation_colors <- list(
  Recurrence = c("NoRecurrence" = "#b2df8a", "Recurrence" = "#f81f2C")
)

pheatmap(
  mat_top20,
  scale = "row",
  clustering_distance_cols = "euclidean",
  clustering_distance_rows = "euclidean",
  annotation_col = ann,
  show_colnames = FALSE,
  fontsize_row = 8,
  annotation_colors = annotation_colors,
  main = "Figure 4.2 Top 20 Recurrence-Associated Genes Heatmap"
)
```
## Figure 4.2 Heatmap top20

```{r Figure 4.3 volcano plot}
res_df <- as.data.frame(res_deseq)
res_df$gene <- rownames(res_df)
res_df$ensembl_clean <- sub("\\..*", "", res_df$gene)

# mapping table for topN:
top_map <- topN %>%
  select(ensembl_clean, label) %>%
  rename(GeneLabel = label)

res_df <- res_df %>%
  left_join(top_map, by = "ensembl_clean") %>%
  mutate(
    highlight = ifelse(is.na(GeneLabel), "Other", "Top20"),
    label     = GeneLabel    
  )
#table(res_df$highlight)

ggplot(res_df, aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(aes(color = highlight), alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype="dashed") +
  geom_vline(xintercept = c(-1, 1), linetype="dotted") +
  scale_color_manual(values = c(Other = "grey50", Top20 = "red")) +
  geom_text_repel(
    data = subset(res_df, highlight == "Top20"),
    aes(label = label),
    size = 3,
    max.overlaps = Inf
  ) +
  theme_bw(base_size = 14) +
  labs(
    title = "Figure 4.3 Top 20 Recurrence-Associate Genes Volcano Plot",
    x = "log2 Fold Change",
    y = "-log10(p-value)"
  )
```
## Figure 4.3 volcano plot


```{r Figures 4.4-8 Boxplots}
gene1 <- "ENSG00000011677"  # GABRA3
df_plot1 <- data.frame(
  expr_raw = norm_counts[gene1, clin_rec$PatientID],
  Recurrence = clin_rec$Recurrence
)
df_plot1$expr_log1p <- log1p(df_plot1$expr_raw)
y_max <- quantile(df_plot1$expr_log1p, 0.99, na.rm = TRUE)
ggplot(df_plot1, aes(x = Recurrence, y = expr_log1p, fill = Recurrence)) +
  geom_boxplot(outlier.alpha = 0.4) +
  scale_fill_manual(values = c(NoRecurrence = "#b2df8a", Recurrence = "#f81f2C")) +
  coord_cartesian(ylim = c(0, y_max)) +
  labs(
    title = "Figure 4.4 Expression of GABRA3",
    x = "Recurrence Status",
    y = "Normalized expression (log1p counts)"
  ) +
  theme_bw(base_size = 14) +
  theme(
    legend.position = "right",
    plot.title = element_text(hjust = 0.5)
  )

gene2 <- "ENSG00000075618"  # FSCN1 
df_plot2 <- data.frame(
  expr_raw = norm_counts[gene2, clin_rec$PatientID],
  Recurrence = clin_rec$Recurrence
)
df_plot2$expr_log1p <- log1p(df_plot2$expr_raw)
y_max <- quantile(df_plot2$expr_log1p, 0.99, na.rm = TRUE)
ggplot(df_plot2, aes(x = Recurrence, y = expr_log1p, fill = Recurrence)) +
  geom_boxplot(outlier.alpha = 0.4) +
  scale_fill_manual(values = c(NoRecurrence = "#b2df8a", Recurrence = "#f81f2C")) +
  coord_cartesian(ylim = c(0, y_max)) +
  labs(
    title = "Figure 4.5 Expression of FSCN1",
    x = "Recurrence Status",
    y = "Normalized expression (log1p counts)"
  ) +
  theme_bw(base_size = 14) +
  theme(
    legend.position = "right",
    plot.title = element_text(hjust = 0.5)
  )

gene3 <- "ENSG00000091138"  # SLC26A3 
df_plot3 <- data.frame(
  expr_raw = norm_counts[gene3, clin_rec$PatientID],
  Recurrence = clin_rec$Recurrence
)
df_plot3$expr_log1p <- log1p(df_plot3$expr_raw)
y_max <- quantile(df_plot3$expr_log1p, 0.99, na.rm = TRUE)
ggplot(df_plot3, aes(x = Recurrence, y = expr_log1p, fill = Recurrence)) +
  geom_boxplot(outlier.alpha = 0.4) +
  scale_fill_manual(values = c(NoRecurrence = "#b2df8a", Recurrence = "#f81f2C")) +
  coord_cartesian(ylim = c(0, y_max)) +
  labs(
    title = "Figure 4.6 Expression of SLC26A3",
    x = "Recurrence Status",
    y = "Normalized expression (log1p counts)"
  ) +
  theme_bw(base_size = 14) +
  theme(
    legend.position = "right",
    plot.title = element_text(hjust = 0.5)
  )

gene4 <- "ENSG00000092054"  # MYH7 
df_plot4 <- data.frame(
  expr_raw = norm_counts[gene4, clin_rec$PatientID],
  Recurrence = clin_rec$Recurrence
)
df_plot4$expr_log1p <- log1p(df_plot4$expr_raw)
y_max <- quantile(df_plot4$expr_log1p, 0.99, na.rm = TRUE)
ggplot(df_plot4, aes(x = Recurrence, y = expr_log1p, fill = Recurrence)) +
  geom_boxplot(outlier.alpha = 0.4) +
  scale_fill_manual(values = c(NoRecurrence = "#b2df8a", Recurrence = "#f81f2C")) +
  coord_cartesian(ylim = c(0, y_max)) +
  labs(
    title = "Figure 4.7 Expression of MYH7",
    x = "Recurrence Status",
    y = "Normalized expression (log1p counts)"
  ) +
  theme_bw(base_size = 14) +
  theme(
    legend.position = "right",
    plot.title = element_text(hjust = 0.5)
  )

gene5 <- "ENSG00000102683"  # SGCG ENSG
df_plot5 <- data.frame(
  expr_raw = norm_counts[gene5, clin_rec$PatientID],
  Recurrence = clin_rec$Recurrence
)
df_plot5$expr_log1p <- log1p(df_plot5$expr_raw)
y_max <- quantile(df_plot5$expr_log1p, 0.99, na.rm = TRUE)
ggplot(df_plot5, aes(x = Recurrence, y = expr_log1p, fill = Recurrence)) +
  geom_boxplot(outlier.alpha = 0.4) +
  scale_fill_manual(values = c(NoRecurrence = "#b2df8a", Recurrence = "#f81f2C")) +
  coord_cartesian(ylim = c(0, y_max)) +
  labs(
    title = "Figure 4.8 Expression of SGCG",
    x = "Recurrence Status",
    y = "Normalized expression (log1p counts)"
  ) +
  theme_bw(base_size = 14) +
  theme(
    legend.position = "right",
    plot.title = element_text(hjust = 0.5)
  )
```
## Figures 4.4-8 Boxplots

```{r Figure 4.9 PCA plot}
vsd <- vst(dds, blind = FALSE) # variance stabilizing transform
mat_vst <- assay(vsd)
rownames(mat_vst) <- sub("\\..*", "", rownames(mat_vst))

top_ids <- topN$ensembl_clean
#top_ids[!top_ids %in% rownames(mat_vst)] 
mat_vst_top20 <- mat_vst[top_ids, clin_rec$PatientID]

pca <- prcomp(t(mat_vst_top20), scale. = TRUE)
var_expl <- pca$sdev^2 / sum(pca$sdev^2)

pca_df <- data.frame(
  PC1 = pca$x[,1],
  PC2 = pca$x[,2],
  Recurrence = clin_rec$Recurrence
)
# drop extreme PC1 outlier (lower 1%)
cutoff <- quantile(pca_df$PC1, 0.01)
pca_df_plot <- subset(pca_df, PC1 > cutoff)

ggplot(pca_df_plot, aes(x = PC1, y = PC2, color = Recurrence)) +
  geom_point(size = 2.5, alpha = 0.8) +
  stat_ellipse(level = 0.95, linetype = "dashed") +
  scale_color_manual(values = c(NoRecurrence = "#b2df8a", Recurrence = "#f81f2C")) +
  theme_bw(base_size = 14) +
  labs(
    title = "Figure 4.9 PCA of Top 20 Recurrence-Associated Genes",
    x = paste0("PC1 (", round(100 * var_expl[1]), "%)"),
    y = paste0("PC2 (", round(100 * var_expl[2]), "%)")
  ) +
  theme(
    plot.title = element_text(hjust = 0.5),
    legend.position = "right"
  )
```
```{r Fig4.10 LASSO}
mat_all <- counts(dds, normalized = TRUE)
rownames(mat_all) <- sub("\\..*", "", rownames(mat_all))

# keep only patients used in analysis
mat_all <- mat_all[, clin_rec$PatientID]

# top 1700 most variable genes
gene_var <- apply(mat_all, 1, var)
top2000 <- names(sort(gene_var, decreasing = TRUE))[1:2000]

x <- t(mat_all[top2000, ])
y <- ifelse(clin_rec$Recurrence == "Recurrence", 1, 0)

# Run LASSO:
set.seed(123)
cvfit <- cv.glmnet(x, y, family = "binomial")

lasso_genes <- rownames(coef(cvfit, s="lambda.min"))[
  coef(cvfit, s="lambda.min")[,1] != 0
]
lasso_genes

# Evaluate model:
pred <- predict(cvfit, newx = x, s = "lambda.min", type="response")
roc_obj <- roc(y, as.numeric(pred))
ggroc(roc_obj, colour = "#1f78b4", size = 1.2) +
  geom_abline(slope = 1, intercept = 1, linetype = "dashed", color = "grey60") +
  theme_bw(base_size = 14) +
  labs(
    title = "Figure 4.10 LASSO Recurrence Classifier ROC Curve",
    x = "1 - Specificity",
    y = "Sensitivity"
  ) +
  annotate(
    "text",
    x = 0.65, y = 0.1,
    label = paste0("AUC = ", round(auc(roc_obj), 3)),
    size = 5
  )
```

```{r fig4.11 lasso}
mat_all <- counts(dds, normalized = TRUE)
rownames(mat_all) <- sub("\\..*", "", rownames(mat_all))
mat_all <- mat_all[, clin_rec$PatientID]

# Top 20 DE genes
top20_ids <- topN$ensembl_clean
x <- t(mat_all[top20_ids, ]) # expression matrix
y <- ifelse(clin_rec$Recurrence == "Recurrence", 1, 0) # outcome vector

set.seed(123)
cvfit20 <- cv.glmnet(x, y, family = "binomial", alpha = 1)
coef(cvfit20, s = "lambda.min")

pred20 <- predict(cvfit20, newx = x, s = "lambda.min", type = "response")

roc20 <- roc(y, as.numeric(pred20))
auc20 <- auc(roc20)
auc20

ggroc(roc20, colour = "#1f78b4", size = 1.2) +
  geom_abline(slope = 1, intercept = 1, 
              linetype = "dashed", color = "grey50", linewidth = 1) +
  theme_bw(base_size = 14) +
  labs(
    title = "Figure 4.11 LASSO Classifier ROC Curve (Top 20 DE Genes)",
    x = "1 - Specificity",
    y = "Sensitivity"
  ) +
  annotate(
    "text",
    x = 0.65, y = 0.1,
    label = paste0("AUC = ", round(auc20, 3)),
    size = 5
  ) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.grid = element_blank()
  )
```
